### PR TITLE
Add api docs

### PR DESCRIPTION
**Pull request by bdraco**
## What do these changes do?

Add API docs

## Are there changes in behavior for the user?

no

### CODE REVIEW

**Comment by webknjaz** [comment]
@bdraco would you want to integrate `sphinxcontrib.apidoc` instead of authoring these in separate documents? Here's a configuration example: https://github.com/ansible/awx_plugins.interfaces/blob/devel/docs/conf.py#L126-L138. You basically add a dep, add it to the extensions, make a gitignored folder and include that in the index file. And it builds the pages from docstrings.

File: docs/api.rst [hunk 1]
```diff
@@ -0,0 +1,54 @@
+.. _propcache-api:
+
+============
+Reference
+============
```
Comment (webknjaz) [inline docs/api.rst:5]: These should usually match the word length:
```suggestion
=========
Reference
=========
```

File: docs/index.rst [hunk 2]
```diff
@@ -18,6 +18,11 @@ The API is designed to be nearly identical to the built-in ``cached_property`` c
 except for the additional ``under_cached_property`` class which uses ``self._cache``
 instead of ``self.__dict__`` to store the cached values and prevents ``__set__`` from being called.
 
+API documentation
+------------------
```
Comment (webknjaz) [inline docs/index.rst:22]:
```suggestion
-----------------
```

**Comment by webknjaz** [comment]
Oh.. That might be problematic for the C-extensions, though.

**Comment by bdraco** [comment]
> @bdraco would you want to integrate `sphinxcontrib.apidoc` instead of authoring these in separate documents? Here's a configuration example: [ansible/awx_plugins.interfaces@`devel`/docs/conf.py#L126-L138](https://github.com/ansible/awx_plugins.interfaces/blob/devel/docs/conf.py?rgh-link-date=2024-10-06T00%3A13%3A21Z#L126-L138). You basically add a dep, add it to the extensions, make a gitignored folder and include that in the index file. And it builds the pages from docstrings.

> Oh.. That might be problematic for the C-extensions, though.

I think that makes sense if we end up adding more here.  I'm not really expecting this module will be all that active though.

**Comment by webknjaz** [comment]
@bdraco I've enabled the PR builds in RTD, apparently, that checkbox wasn't selected.

**Comment by Dreamsorcerer** [comment]
> @bdraco would you want to integrate `sphinxcontrib.apidoc` instead of authoring these in separate documents? Here's a configuration example: https://github.com/ansible/awx_plugins.interfaces/blob/devel/docs/conf.py#L126-L138. You basically add a dep, add it to the extensions, make a gitignored folder and include that in the index file. And it builds the pages from docstrings.

I've been meaning to ask if there was a reason we've not done that in any of our projects. We often seem to end up with docs getting a little out-of-sync because they're written manually.

**Comment by bdraco** [comment]
I think yarl is a great place to start with this approach as it has a lot of methods and it’s likely to be a lot more active than this lib

**Comment by webknjaz** [comment]
> > @bdraco would you want to integrate `sphinxcontrib.apidoc` instead of authoring these in separate documents? Here's a configuration example: [ansible/awx_plugins.interfaces@`devel`/docs/conf.py#L126-L138](https://github.com/ansible/awx_plugins.interfaces/blob/devel/docs/conf.py?rgh-link-date=2024-10-06T13%3A19%3A07Z#L126-L138). You basically add a dep, add it to the extensions, make a gitignored folder and include that in the index file. And it builds the pages from docstrings.
> 
> I've been meaning to ask if there was a reason we've not done that in any of our projects. We often seem to end up with docs getting a little out-of-sync because they're written manually.

@Dreamsorcerer yeah, I've looked into it a couple of times in the past actually, just forgot :)
The main problem is that the C-extensions aren't inspectable by Sphinx IIRC (there was an upstream bug somewhere). I don't remember if this can be addressed with some stubs like pyi.
Another issue was that Andrew didn't want the docstrings to be big (but I think we last talked about that over 7 years ago). He was also concervative when I brought up the suggestion to auto-include parts README (and other docs that render on GH too) into Sphinx docs to avoid duplication, like I do in all my other projects. I've since seen some mentions that some projects (CPython maybe?) were migrating away from in-docstring docs because it was hurting the readabiliy. But personally, I'm very much in favor of automating stuff.

**Comment by webknjaz** [comment]
> I think yarl is a great place to start with this approach as it has a lot of methods and it’s likely to be a lot more active than this lib

On the contrary, I'd start with a low-traffic lib where experiments could happen in a safe setting, before scaling to places that would require more integration work that might not pay off because of some limitations.

**Pull request merged**